ouroboros-consensus-cardano/tools/db-validator/Main.hs

```diff
@@ -40,25 +32,45 @@ import qualified Ouroboros.Consensus.Storage.ChainDB as ChainDB
 import           Ouroboros.Consensus.Storage.ChainDB.Impl.Args (fromChainDbArgs)
 import qualified Ouroboros.Consensus.Storage.ChainDB.Impl.ImmDB as ImmDB
 
-import           Ouroboros.Consensus.Byron.Ledger (ByronBlock)
-import           Ouroboros.Consensus.Byron.Node
+import           Ouroboros.Consensus.Byron.Node (PBftSignatureThreshold (..))
 
-import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock)
-import           Ouroboros.Consensus.Shelley.Node (Nonce (..), ShelleyGenesis,
-                     protocolInfoShelley)
-import           Ouroboros.Consensus.Shelley.Protocol.Crypto
+import           Ouroboros.Consensus.Shelley.Node (Nonce (..))
 
-import           Ouroboros.Consensus.Cardano.Block (CardanoBlock)
-import           Ouroboros.Consensus.Cardano.Node (TriggerHardFork (..),
-                     protocolInfoCardano)
+import           Analysis
+import           Class
+
+main :: IO ()
+main = do
+    cmdLine <- getCmdLine
+    warnUnusedFlags cmdLine
+    case blockType cmdLine of
+      ByronBlock   args -> validateOrAnalyse cmdLine args
+      ShelleyBlock args -> validateOrAnalyse cmdLine args
+      CardanoBlock args -> validateOrAnalyse cmdLine args
+
+warnUnusedFlags :: CmdLine -> IO ()
+warnUnusedFlags CmdLine{..} = case (analysis, onlyImmDB) of
+  (Just _, True)
+    -> putStrLn "Warning: Flag onlyImmDB is ignored on analysis mode"
+  _ -> return ()
```

Comment:
Actually, couldn't we just let analyses take this flag into account too? In other words, analyse the ImmDB or the ChainDB?

Reply:
True. I've been working on extending the analyser for the whole chainDB and make this flag useful.

Reply:
Do you intend to include that in this PR before it's merged?

Reply:
Idea: if we can make `analyse` work with the ChainDB too, then validation would be just another type of analysis, a no-op analysis. This means we can truly merge `analyse` and `validate`! In which case, I would call the tool `db-analyser` instead of `db-validator`. Great, let's do this, this perfectly matches the title of this PR :slightly_smiling_face:.